File: test/nn.py
from engine import Value
import random

class Neuron:
    def __init__(self, nin):
        self.weights = [Value(random.uniform(-1.0, 1.0)) for i in range(nin)]
        self.bias = Value(random.uniform(-1.0, 1.0))

    def __call__(self, x): # forward pass
        _sum = self.bias
        for xi, wi in zip(x, self.weights):
            _sum = _sum + (xi * wi)
        return _sum.tanh()

    def parameters(self):
        return self.weights + [self.bias]

class Layer: 
    def __init__(self, nin, nout):
        self.neurons = [Neuron(nin) for i in range(nout)]

    def __call__(self, x): # x is nin
        out = [neuron(x) for neuron in self.neurons]
        return out[0] if len(out) == 1 else out
    
    def parameters(self):
        params = []
        for neuron in self.neurons:
            ps = neuron.parameters()
            params.extend(ps)
        return params
        # return [p for neuron in self.neurons for p in neuron.parameters()]

class MLP: 
    def __init__(self, nin, nouts):
        self.layers = []
        prev = nin
        for i in nouts:
            self.layers.append(Layer(prev, i))
            prev = i

    def __call__(self, x): # forward pass
        y = x
        for i in self.layers:
            y = i(y)
        return y
    
    def parameters(self):
        params = []
        for layer in self.layers:
            ps = layer.parameters()
            params.extend(ps)
        return params

File: benchmarks/nets/nn.cpp
#include "nn.h"
#include <random>
#include <utility>

Neuron::Neuron(int nin) {
  std::random_device seed;
  std::mt19937 gen{seed()};
  std::uniform_real_distribution<> dist{-1.0, 1.0};
  for (int i = 0; i < nin; i++) {
    this->weights_.emplace_back(dist(gen));
  }
  this->bias_ = Variable(dist(gen));
}

std::vector<Variable> Neuron::parameters() {
  std::vector<Variable> ret(this->weights_.begin(), this->weights_.end());
  ret.push_back(this->bias_);
  return ret;
}

Variable Neuron::operator()(const std::vector<Variable>& x) {
  Variable ret = Variable(this->bias_.data());
  for (int i = 0; i < static_cast<int>(x.size()); i++) {
    ret = ret + (this->weights_[i] * x[i]);
  }
  return ret;
}

Layer::Layer(int nin, int nout) {
  for (int i = 0; i < nout; i++) {
    this->neurons_.push_back(std::make_shared<Neuron>(nin));
  }
}

std::vector<Variable> Layer::parameters() {
  std::vector<Variable> params;
  for (auto& neuron : this->neurons_) {
    std::vector<Variable> n_params = neuron->parameters();
    params.insert(params.end(), n_params.begin(), n_params.end());
  }
  return params;
}

std::vector<Variable> Layer::operator()(const std::vector<Variable>& x,
                                        bool activ) {
  std::vector<Variable> ret;
  for (auto& neuron : this->neurons_) {
    Variable val = (*neuron)(x);
    if (activ) {
      ret.push_back(val.relu());
    } else {
      ret.push_back(val);
    }
  }
  return ret;
}

MultiLayerPerceptron::MultiLayerPerceptron(int nin,
                                           const std::vector<int>& nouts) {
  int prev = nin;
  for (int nout : nouts) {
    this->layers_.push_back(std::make_shared<Layer>(prev, nout));
    prev = nout;
  }
}

std::vector<Variable> MultiLayerPerceptron::parameters() {
  std::vector<Variable> params;
  for (auto& layer : this->layers_) {
    std::vector<Variable> l_params = layer->parameters();
    params.insert(params.end(), l_params.begin(), l_params.end());
  }
  return params;
}

std::vector<Variable> MultiLayerPerceptron::operator()(
    const std::vector<Variable>& x) {
  std::vector<Variable> ret = std::move(x);
  for (auto& layer : this->layers_) {
    ret = (*layer)(ret);
  }
  return ret;
}